### PR TITLE
python/matching-brackets: new mentoring notes

### PR DESCRIPTION
**Pull request by IsaacG**
@BethanyG  for review

### CODE REVIEW

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            unmatched.append(character)
+        if character in matching:
+            if not unmatched:
+                return False
+            # Peek at the current context
+            if unmatched[-1] == matching[character]:
+                # Pop a context
+                unmatched.pop()
+            else:
+                return False
```
Comment (roxgib) [inline tracks/python/exercises/matching-brackets/mentoring.md:25]: Why is the peak necessary? 
```
if unmatched.pop() != matching[character]:
    return False
```
Short circuit evaluation could also be used here, though that might be pushing it for an 'easy' task?
```
if not unmatched or unmatched.pop() != matching[character]:
          return False
```

**Review by roxgib — CHANGES_REQUESTED** [review]

File: tracks/python/exercises/matching-brackets/mentoring.md (hunk 1 above)
Comment (IsaacG) [reply tracks/python/exercises/matching-brackets/mentoring.md:25]: Peaking isn't *necessary*. But it's a good solution.

Conceptually:
1. a pop occurs when a match is found. If there's no match, the value should not be popped. This allows, say, the remaining stack to be accurately printed or whatnot.
2. this exercise is all about using a stack and peak is part of the standard stack operations.

**Comment by IsaacG** [comment]
@BethanyG Ping?

**Comment by IsaacG** [comment]
@ErikSchierboom mind giving this a review/merge?

**Comment by BethanyG** [comment]
@IsaacG --  I don't know why this is, but if you don't request me as a reviewer, the ping only goes to GH notifications and not my inbox.  I've been tied up with other projects, so did not login to GH / notice this.  I'll review shortly.

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    return not unmatched
+```
+
+## Talking Points
+
+### Maintaining Context
+
+The key to solving this problem is recognizing that the solution is all about maintaining contexts.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:33]:
```suggestion
The key to solving this problem is recognizing that the solution is all about _maintaining context_.
```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+When processing an open bracket, the context changes to that bracket type, adding a new context.
+Prior open brackets need to be preserved, but we do not need to access them until the new context is closed.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:35]:
```suggestion
Prior open brackets need to be preserved, but do not need to be accessed until the current context is closed.
```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+When deciding is a given close bracket is valid, the current context is all that is needed - what the last open bracket was.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:36]:
```suggestion
When deciding if a given close bracket is valid, the current context (the last known open bracket) is all that is needed.
```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+If the current context matches the close bracket, the context is closed.
+If the current context does not match, the close bracket is invalid.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:38]:
```suggestion
If the current context does **not** match, the close bracket is invalid.
```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+A stack provides a method for maintaining context: `stack.push()` to add a new context; `stack.pop()` to remove a context; and `stack.peek()` to check the current context.
+Lastly, `len(stack)` is needed to check if there is any remaining context once the string is finished.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:41]:
```suggestion
A stack provides a method for maintaining context: 
- `stack.push()` to add a new context.
- `stack.pop()` to remove a context.
- `stack.peek()` to check the current context.
- Lastly, `len(stack)` to check if there is any remaining context once the string is finished.

```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+Python's list is a superset of a stack, with it's `list.append()`, `list.pop()`, `list[-1]`, and `len(list)`.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:43]:
```suggestion
Python's [list][list-python] data structure is a superset of a stack, with methods `list.append()`, `list.pop()`, `list[-1]` (_indexing_), and `len(list)`.
```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+stack:      # pop("]")
+stack: (    # push("(")
+stack:      # pop(")")
+stack: [    # push("[")
+stack:      # pop("]")
+```
+
+### Regular Expressions: Language Theory
+
+Note: This section is heavy on the theory and light on details.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:63]:
```suggestion
**Note:** This section is heavy on theory and light on implementation details.
```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+The intent is to provide direction if one would like to delve deeper into this topic.
+
+Regular expressions are often a go-to for parsing strings.
+However, regular expressions are a poor fit for this particular exercise.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:67]:
```suggestion
However, standard regular expressions are a poor fit for this particular exercise.
```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+Below is some theory which explains why it is a poor fit.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:68]:
```suggestion
Below is some theory which explains the issue.
```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+In formal language theory, computer science and linguistics, the Chomsky hierarchy [Chomsky hierarchy][chomsky-hierarchy] is a hierarchy of *formal grammars*.
+A grammar is a tool used to describe a *language*, such as mathematical expressions, a programming language or an HTML document.
+
+Regular expressions provide a way to describe a "[regular language][regular-language]".
+Regular languages are [context free][context-free], which means they cannot maintain a current context or operate based on contexts.
+Languages which use a stack as called [context sensitive][context-sensitive] and *can* maintain a context.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:75]:
```suggestion
Languages which use a stack are called [context sensitive][context-sensitive] languages and *can* maintain a context.
```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+Languages such as simple arithmetic expressions (e.g. `x = a + m * b`) are regular and can be matched using regular expressions.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:77]:
```suggestion
Simple arithmetic expressions (e.g. `x = a + m * b`) are considered regular languages and can be matched using regular expressions.
```

File: tracks/python/exercises/matching-brackets/mentoring.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+Languages such as HTML - or bracket matching, require a context sensitive grammer and cannot be matched using regular expressions.
```
Comment (BethanyG) [inline tracks/python/exercises/matching-brackets/mentoring.md:78]:
```suggestion
However, languages such as HTML - or bracket matching, require a context sensitive grammar and **cannot** be matched using regular expressions.
```

**Review by BethanyG — APPROVED** [review]
All of these are suggestions, so I am marking this as approved.  However, there are a few spelling and grammar corrections, which I recommend accepting.

**Comment by ErikSchierboom** [comment]
@IsaacG As bethany has approved, I'll leave the rest (possibly applying suggestions and/or merging) to you two :)

**Comment by BethanyG** [comment]
@IsaacG -- I am going to leave the accept/reject edits to you..and you can merge when ready.  😄

**Comment by IsaacG** [comment]
Thanks, @ErikSchierboom
Thank you for the feedback, @BethanyG. Applied. Please take another look.

**Review by BethanyG — APPROVED** [review]

File: tracks/python/exercises/matching-brackets/mentoring.md (hunk 1 above)
Comment (roxgib) [reply tracks/python/exercises/matching-brackets/mentoring.md:25]: Okay, makes sense!

**Pull request merged**